subsys/bluetooth/host/l2cap.c

```diff
@@ -1086,6 +1084,10 @@ static uint16_t l2cap_chan_accept(struct bt_conn *conn,
 		(*chan)->ops->connected(*chan);
 	}
 
+	/* Give credits */
+	l2cap_chan_tx_give_credits(le_chan, credits);
+	l2cap_chan_rx_give_credits(le_chan, le_chan->rx.init_credits);
```

Comment:
Since these are moved to after the `connected` callback, doesn't this prevent the user to use the channel in the `connected` callback immediately? i.e. if you call `bt_l2cap_chan_send` in the `connected` callback, that would now hit

```c
	if (!test_and_dec(&ch->tx.credits)) {
		LOG_WRN("No credits to transmit packet");
		return -EAGAIN;
	}
```

in `l2cap_chan_le_send`, and return an error. 

This change then changes the behavior of the L2CAP API, and that is something we might want to avoid.

Reply:
When going through `bt_l2cap_chan_send_cb` the fragment below should handle this.

	/* Queue if there are pending segments left from previous packet or
	 * there are no credits available.
	 */
	if (le_chan->tx_buf || !k_fifo_is_empty(&le_chan->tx_queue) ||
	    !atomic_get(&le_chan->tx.credits)) {
		l2cap_tx_meta_data(buf)->sent = 0;
		net_buf_put(&le_chan->tx_queue, buf);
		k_work_submit(&le_chan->tx_work);
		return 0;
	}

When going through `l2cap_chan_tx_process` the fragment below will reschedule:

		sent = l2cap_chan_le_send_sdu(ch, &buf, sent);
		if (sent < 0) {
			if (sent == -EAGAIN) {
				ch->tx_buf = buf;
				/* If we don't reschedule, and the app doesn't nudge l2cap (e.g. by
				 * sending another SDU), the channel will be stuck in limbo. To
				 * prevent this, we attempt to re-schedule the work item for every
				 * channel on every connection when an SDU has successfully been
				 * sent.
				 */
			} else {
				net_buf_unref(buf);
			}
			break;
		}

Without the change, the peer will hit the condition below, which means that anything sent in the `connected` callbacks will be lost anyway.

	if (!test_and_dec(&chan->rx.credits)) {
		LOG_ERR("No credits to receive packet");
		bt_l2cap_chan_disconnect(&chan->chan);
		return;
	}

Reply:
We have received some initial credits in the L2CAP connection request. We should be able to immediately send something using those credits. We should also be able to receive something immediately after receiving the L2CAP connect response since we gave some credits in the connection request.

Reply:
Thanks; I missed the `k_work_submit(&le_chan->tx_work);` if no credits are available :)